share/wake/lib/core/map.wake

```diff
@@ -0,0 +1,548 @@
+#   ```
+export def munionWith (fn: k => v => v => v) (left: Map k v) (right: Map k v): Map k v =
+    def Map _ leftTree = left
+    # tunionWith uses the comparison function of the right tree, so similarly
+    # attach the raw key comparison function of the right map to the output
+    def Map cmpKey rightTree = right
+    def pairFn l r =
+        def Pair _ lv = l
+        def Pair rk rv = r
+        rk :-> fn rk lv rv
+    tunionWith pairFn leftTree rightTree
+    | Map cmpKey
+
+# Remove all keys from the left map which occur (regardless of value) in the right.
+# If the key comparison function differs between the two inputs, then the one
+# from the `right` is used.
+#
+# *Examples:*
+#   ```
+#   def left  = listToMap scmp ("a" → 1, "b" → 2, Nil)
+#   def right = listToMap scmp ("b" → 11, "f" → 15, Nil)
+#
+#   msubtract left right | mlookup "a" = Some 1
+#   msubtract left right | mlookup "f" = None
+#   msubtract left right | mlookup "b" = None
+#   ```
+export def msubtract (left: Map k v) (right: Map k v): Map k v =
+    def Map _ leftTree = left
+    # tsubtract uses the comparison function of the right tree, so similarly
+    # attach the raw key comparison function of the right map to the output
+    def Map cmpKey rightTree = right
+    tsubtract leftTree rightTree
+    | Map cmpKey
+
+# Remove all keys from the left map which do not occur in the right.
+# For all keys, the value from `left` is kept and the one from `right` is
+# discarded.  However, if the key comparison function differs between the two
+# inputs, then the one from the `right` is used.
+#
+# For a similar function which preserves the original value contained in the
+# map, see `mintersectWith`.
+#
+# *Examples:*
+#   ```
+#   def left  = listToMap scmp ("a" → 1, "b" → 2, Nil)
+#   def right = listToMap scmp ("b" → 11, "f" → 15, Nil)
+#
+#   mintersect left right | mlookup "a" = None
+#   mintersect left right | mlookup "f" = None
+#   mintersect left right | mlookup "b" = Some 2
+#   ```
+export def mintersect (left: Map k v) (right: Map k v): Map k v =
+    def Map _ leftTree = left
+    # tintersect uses the comparison function of the right tree, so similarly
+    # attach the raw key comparison function of the right map to the output
+    def Map cmpKey rightTree = right
+    tintersect leftTree rightTree
+    | Map cmpKey
+
+# Remove all keys from the left map which do not occur in the right, joining values.
+# If the key comparison function differs between the two inputs, then the one
+# from the `right` is used.
+#
+# *Examples:*
+#   ```
+#   def left  = listToMap scmp ("a" → 1, "b" → 2, Nil)
+#   def right = listToMap scmp ("b" → 11, "f" → 15, Nil)
+#
+#   mintersectWith left right | mlookup "a" = None
+#   mintersectWith left right | mlookup "f" = None
+#   mintersectWith left right | mlookup "b" = Some 13
+#   ```
+export def mintersectWith (fn: k => v => v => v) (left: Map k v) (right: Map k v): Map k v =
```

Comment:
The value type should be allowed to change to a v2 here I think?

Reply:
This one could indeed become a `v2`, though.  My only argument would be for consistency with the others, but maybe the abstraction is better.

Reply:
I don't see a lot of use cases for being more abstract but I figure why limit it here. Up to you, minor issue.